docs/spec.md

```diff
@@ -2288,23 +2288,59 @@ Computes dot products between slices of `lhs` and slices of `rhs` and produces a
+            storage_type=storage_type(rhs),
+            storage_min=storage_min(rhs),
+            storage_max=storage_max(rhs),
+            quantization_dimension=lhs_batching_dimensions[0],
+            scale=scale, zero_point=zero_point)
+  ```
+
+* `If lhs uses quantized tensor or rhs uses non-quantized tensor:`
+  * `effective_lhs = lhs`.
+* `lhs_result_dimensions = [d for d in axes(effective_lhs) and d not in lhs_batching_dimensions and d not in lhs_contracting_dimensions]`.
 * `rhs_result_dimensions = [d for d in axes(rhs) and d not in rhs_batching_dimensions and d not in rhs_contracting_dimensions]`.
 * `result_batching_index + result_lhs_index + result_rhs_index = result_index`
   where `size(result_batching_index) = size(lhs_batching_dimensions)`,
   `size(result_lhs_index) = size(lhs_result_dimensions)` and
   `size(result_rhs_index) = size(rhs_result_dimensions)`.
-* `transposed_lhs = transpose(lhs, lhs_batching_dimensions + lhs_result_dimensions + lhs_contracting_dimensions)`.
-* `transposed_lhs_slice = slice(result_batching_index + result_lhs_index + [:, ..., :])`.
-* `reshaped_lhs_slice = reshape(transposed_lhs_slice, dims(lhs, lhs_contracting_dimensions))`.
+* `transposed_lhs = transpose(effective_lhs, lhs_batching_dimensions + lhs_result_dimensions + lhs_contracting_dimensions)`.
+* `transposed_lhs_slice = slice(transposed_lhs, result_batching_index + result_lhs_index + [:, ..., :])`.
+* `reshaped_lhs_slice = reshape(transposed_lhs_slice, dims(effective_lhs, lhs_contracting_dimensions))`.
 * `transposed_rhs = transpose(rhs, rhs_batching_dimensions + rhs_result_dimensions + rhs_contracting_dimensions)`.
-* `transposed_rhs_slice = slice(result_batching_index + result_rhs_index + [:, ..., :])`.
+* `transposed_rhs_slice = slice(transposed_rhs, result_batching_index + result_rhs_index + [:, ..., :])`.
 * `reshaped_rhs_slice = reshape(transposed_rhs_slice, dims(rhs, rhs_contracting_dimensions))`.
-* `dot_product = reduce(
-  inputs=[multiply(reshaped_lhs_slice, reshaped_rhs_slice)],
-  init_values=[0],
-  dimensions=[0, ..., size(lhs_contracting_dimensions) - 1],
-  body=lambda x, y: add(x, y))`.
+* For `is_non_quantized_tensor(lhs) and is_non_quantized_tensor(rhs)`:
+  * `dot_product = reduce(
+      inputs=[multiply(reshaped_lhs_slice, reshaped_rhs_slice)],
+      init_values=[0],
+      dimensions=[0, ..., size(lhs_contracting_dimensions) - 1],
+      body=lambda x, y: add(x, y))`.
+* For `is_quantized_tensor(lhs) and is_quantized_tensor(rhs)`:
+  * `integer_dot_product = reduce(
+      inputs=[multiply((reshaped_lhs_slice - zero_point(reshaped_lhs_slice)),
+                       (reshaped_rhs_slice - zero_point(reshaped_rhs_slice))],
+      init_values=[0],
+      dimensions=[0, ..., size(lhs_contracting_dimensions) - 1],
+      body=lambda x, y: add(x, y))`.
+  * `rounded_dot_product = round_nearest_even(integer_dot_product * (scale(reshaped_lhs_slice) * scale(reshape_rhs_slice) / scale(result)))`.
```

Comment:
The scaling on this line is defined in terms of floating-point scale values and so can be problematic for integer only processing. Similar to the discussion on Add quantization, can this scaling be converted to an integer multiply and shift rescale for integer processing?”

Reply:
Hi @dominicsymes 

>> can this scaling be converted to an integer multiply and shift rescale for integer processing?”

The current proposal uses the [textbook](https://arxiv.org/pdf/1712.05877.pdf) [Section 2.2] approach to define the specification. Specifically to ensure  bit-exactness, we can have the following arrangement: 

Following the ongoing discussion in https://groups.google.com/a/openxla.org/g/openxla-discuss/c/iwE9is49SS4, here is the rough sketch of the proposal that we can plan for:

1. Other than the uniform quantized type with floating-point scale, StableHLO to have a type with integer [multiplier, shift]. 
2. There will be defined conversion between the the two types. 
3. For the sake generality of the specification, I would like to propose that the consumers of StableHLO to be given the liberty to generate rescale operation, normalizing the operands of any particular op, in their preferred way. Given that, even for for particular add op, there could be different options of doing the rescale, I think these rescaling choices to be better governed and applied by the consumers only.

Happy to understand your perceptive on this.

Reply:
Hi @sdasgup3

Thanks for describing the current proposal. The general scheme seems fine to me. In terms of bit-exactness, I think it would be useful to have a reference conversion to integer rescale operations. This could then be implemented bit-exactly or used as a comparison point if a consumer varies how the rescale operations are generated as you propose in (3).

Reply:
Hi @dominicsymes 

The usefulness of a reference conversion to rescale operations makes perfect sense to me. I believe that could be realized using a legalization pass from StableHLO, with int [miltiplier,shift] parameters, to TOSA rescale operators, similar to the one we have in [legalize_tfl.cc](https://github.com/tensorflow/tensorflow/blob/60f7a770d64cb6cb3a93f84c291272ec51304d31/tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc#L1). 

Let me get back you with more information on addressing this.

Reply:
Hi @dominicsymes 
I have updated the issue #1404 with  https://github.com/openxla/stablehlo/issues/1404#issuecomment-1539225320 to track progress on the proposed reference converter.